config/configcompression/compressiontype.go

```diff
@@ -3,7 +3,13 @@
 
 package configcompression // import "go.opentelemetry.io/collector/config/configcompression"
 
-import "fmt"
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/klauspost/compress/zlib"
+)
 
 // Type represents a compression method
 type Type string
```

Comment:
I don't think we can reuse this type and keep the level encoded in the string. API users may rely on `configcompression.Type == configcompression.TypeGzip` which won't work anymore...

I think we need new additional types like these

```
type Config struct {
	Type Type
	Level Level
}

type Level int
```

`Config` may be also called `TypeWithLevel`.

Given that this module is 1.x, it should be okay to add. However, we would need to make a breaking change to the Co API (not an end-user breaking change) in config HTTP by changing the `confighttp.ClientConfig.Compression` field from `compressionconfig.Type` to `compressionconfig.Config`, which should be okay given that the confighttp module is still 0.x.

Then we can have all validation in `Config.UnmarshalText` and assume that Config is always valid

Reply:
ok this is kind of what I had before in an earlier commit - https://github.com/open-telemetry/opentelemetry-collector/pull/11084/commits/74b1d10754ea0d5f128241058b847c1842ae9ab3#diff-a559328153338200c250ddd678e69c4d0576cad113fa1feaed06b65548e33752R28

It sounds like I should go back to that approach?

Reply:
> confighttp.ClientConfig.Compression field from compressionconfig.Type to compressionconfig.Config
won't this be a breaking change for the user as well?, i.e with the above suggested change we would need two fields to specify compression like below

```
compression:
  type: "gzip"
  level: 1
```

Reply:
@dmitryax I've made changes based on the above comment. Please let me know if this is what you had in mind.

Reply:
looks like the tests are failing because of errors like below, since this is now a breaking change

`Error: config/config.go:147:24: invalid operation: cfg.Compression != "" (mismatched types configcompression.TypeWithLevel and untyped string)`

Reply:
> won't this be a breaking change for the user as well?, 

No, we should keep the string encoding `<type>/<level>` where the `/<level>` part is optional

Reply:
ok, I guess what I don't understand is how this is possible if we change the type from `compressionconfig.Type` to `compressionconfig.Config`, the latter being a struct. Doesn't this imply that we're moving away from `<type>/<level>`?

Reply:
I may have misunderstood the ask here. The following has been implemented

- Keep the compressionType
- Define new TypeWithLevel
- Unmarshal the `<type>/<level>` string encoding to set type and level in `TypeWithLevel`. Also add validation logic in here

Reply:
> No, we should keep the string encoding <type>/<level> where the /<level> part is optional

The latest commit ensures that the `/level` part is optional

Reply:
I have an alternative proposal - what do you think of the below @dmitryax?

---

Users who want to specify a level should use structured config like:

```yaml
compression:
  type: gzip
  level: 9
```

For backwards compatibility support parsing the algorithm name ("type"), but don't invent a new format for including the level in there. i.e. you could specify `compression: gzip` (implies using default compression level), but not `compression: gzip/9` (if you need that, use the structured approach).

To support the above, move validation to its own Validate method, so it applies to both the backwards compatible string parsing case & structured config case.

Reply:
@axw, are you suggesting keeping the mixed type value: string for plain type and struct for type+level? I'm not sure I like this approach. We don't have anything like this in the collector config interface. I think we should keep it strictly typed. I'm actually surprised that it currently accepts both types, that was not desirable.

What if we just add another extra optional field called `compression_params` or `compression_config`  _in addition_ to `compression`? It'll be a fully backward-compatible additive change. It'll also allow us to add any other compression related configs in the future. To define compression level the users would need to provide:
```
compression: gzip
compression_params:
  level: 9
```

@rnishtala-sumo @axw @andrzej-stencel @codeboten WDYT?

Reply:
> @axw, are you suggesting keeping the mixed type value: string for plain type and struct for type+level?

Yes, but only for backwards compatibility. I would suggest phasing out the plain string format, leaving only the struct format.

> What if we just add another extra optional field called compression_params or compression_config in addition to compression? It'll be a fully backward-compatible additive change. It'll also allow us to add any other compression related configs in the future. To define compression level the users would need to provide:

I prefer this over adding the level to the string format, but it means the configuration is no longer a single object/type. So this may lead to more cumbersome code.

Reply:
This is an interesting approach to handle backward compatibility for such change. I'm open to this. Interested to hear other opinions from other @open-telemetry/collector-approvers

Reply:
Just as a reminder, we started of including the level as part of the string, to keep this a non-breaking change for the user. I'd personally like to keep it this way, since changing this from a `string` to a `struct` is a major breaking change to a commonly used feature.

Maybe we keep `compression` a string (with the optional level) and introduce a new field in future (`compression_config`) that is a struct from the beginning which could allow for more compression config params. 

Example:

```
compression_config:
  type: gzip
  level: 1
```

This approach allows us to phase out (deprecate and remove) `compression` after introducing `compression_config`. I think this is less confusing for a user, since they'd be clear about the type of the `compression` field.

Reply:
@rnishtala-sumo, @axw suggests an approach on how to convert the field from string to struct in a backward compatible way.

Reply:
> To support the above, move validation to its own Validate method, so it applies to both the backwards compatible string parsing case & structured config case

right, I got that. Just thought using the same field as a string and as a struct might be confusing to the user, until we phase out the string type. Did we follow this pattern elsewhere?

Reply:
> Did we follow this pattern elsewhere?

This is a new pattern. We haven't considered it yet. But moving from string to struct is something that we potentially can run into with other config fields. So if there is a solution for that, it would be interesting to try.

Reply:
ok, I think all that's needed to do this is to keep unmarshal as is and move the validation to a new method.

Reply:
Configuring compression type is probably a (reasonably) widely used feature. Configuring compression details on the other hand will probably only be used by a small number of users. I'm not sure if we want to introduce a breaking change (even if with a long deprecation time) for a big number of users to satisfy needs of a small number of users?

On the other hand, after giving it a bit of thought, the "stringly typed" approach is also not ideal. I would lean towards the approach with leaving `compression` a string as it is now and adding `compression_config` or similar field for additional settings. Unless we feel it's too awkward and the breaking change is justified.

Reply:
I have pushed the changes suggested by @axw in this [commit](https://github.com/open-telemetry/opentelemetry-collector/pull/11084/commits/976f44c646fe92c5909f9780df3a43a3565aefd2). This does the following:
- Moves the level validation outside Unmarshal
- Allows the compression to be specified as a string/struct

Reply:
JFYI, I tested that with the above change both of the following configuration are valid

compression struct
```
exporters:
  otlphttp:
    endpoint: otelcol:4317
    compression:
      type: gzip
      level: 1
```

compression string
```
exporters:
  otlphttp:
    endpoint: otelcol:4317
    compression: gzip
```

Reply:
Thanks @rnishtala-sumo. To be explicit, I (obviously) think this approach looks good :)

I suppose we'll need to wait for a couple more opinions from approvers to make a final decision

Reply:
@dmitryax @andrzej-stencel do you think we could proceed with @axw's approach?

Reply:
Update: We discussed this during the Collector SIG call on Dec 4. Most of the Collector leads present there (@mx-psi and @jpkrohling) voted to introduce another field (`compression_params`) to avoid breaking the existing user interface even in a backward compatible way. @rnishtala-sumo is going to submit another PR to implement that approach.

Reply:
Here's the new PR with option2 - https://github.com/open-telemetry/opentelemetry-collector/pull/11805